In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 1. LÀM SẠCH DỮ LIỆU

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df_12_baoluu = pd.read_csv("/content/drive/MyDrive/Khai thác dữ liệu và ứng dụng - CS313.N21/Education dataset/12.baoluu.csv")

In [ ]:
df_12_baoluu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1880 entries, 0 to 1879
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          1878 non-null   float64
 1    masv       1878 non-null   object 
 2    tinhtrang  1878 non-null   float64
 3    lydo       1878 non-null   object 
 4    hocky      1878 non-null   float64
 5    namhoc     1878 non-null   float64
 6    soqd       1878 non-null   object 
 7    ngayqd     1878 non-null   object 
dtypes: float64(4), object(4)
memory usage: 117.6+ KB


In [ ]:
# Xóa cột id
df_12_baoluu = df_12_baoluu.iloc[:, 1:]

# Chuẩn hóa tên cột (bỏ khoảng trắng trước tên cột)
df_12_baoluu = df_12_baoluu.rename(columns = {' masv': 'masv',
                                              ' tinhtrang': 'tinhtrang',
                                              ' lydo': 'lydo',
                                              ' hocky': 'hocky',
                                              ' namhoc':'namhoc',
                                              ' soqd': 'soqd',
                                              ' ngayqd': 'ngayqd'})
# Kiểm tra
df_12_baoluu.columns

Index(['masv', 'tinhtrang', 'lydo', 'hocky', 'namhoc', 'soqd', 'ngayqd'], dtype='object')

In [ ]:
# Xóa 2 dòng không chứa dữ liệu
df_12_baoluu.dropna(axis=0, inplace=True)

In [ ]:
# Chuyển kiểu dữ liệu của các cột tinhtrang, hocky, namhoc từ float64 sang int
df_12_baoluu['tinhtrang'] = df_12_baoluu['tinhtrang'].astype(int)
df_12_baoluu['hocky'] = df_12_baoluu['hocky'].astype(int)
df_12_baoluu['namhoc'] = df_12_baoluu['namhoc'].astype(int)

# Kiểm tra
df_12_baoluu.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1878 entries, 0 to 1879
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   masv       1878 non-null   object
 1   tinhtrang  1878 non-null   int64 
 2   lydo       1878 non-null   object
 3   hocky      1878 non-null   int64 
 4   namhoc     1878 non-null   int64 
 5   soqd       1878 non-null   object
 6   ngayqd     1878 non-null   object
dtypes: int64(3), object(4)
memory usage: 117.4+ KB


In [ ]:
df_12_baoluu['lydo'].unique()

array([' TN', ' ',
       ' Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
       ' QĐ điều chỉnh ngành TN', ' Tốt nghiệp'], dtype=object)

In [ ]:
# Chuyển ' ' thành nan
# Bỏ các khoảng trắng đầu chuỗi giá trị

df_12_baoluu.replace(' ', np.nan, inplace=True)
df_12_baoluu.replace(' TN', 'Tốt nghiệp', inplace=True)
df_12_baoluu.replace(' Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
                     'Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
                     inplace=True)
df_12_baoluu.replace(' QĐ điều chỉnh ngành TN', 'QĐ điều chỉnh ngành TN', inplace=True)
df_12_baoluu.replace(' Tốt nghiệp', 'Tốt nghiệp', inplace=True)

# Kiểm tra
df_12_baoluu['lydo'].unique()

array(['Tốt nghiệp', nan,
       'Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
       'QĐ điều chỉnh ngành TN'], dtype=object)

In [ ]:
df_12_baoluu['soqd'].unique()
## Ví soqd với ngayqd có tính chất tương tự nhau nên không cần sử dụng cột soqd
## Suy ra: Không cần chuẩn hóa cột này

array([' 178 /QĐ-ĐHCNTT', ' 713/QD-DHCNTT', ' 713/QĐ-ĐHCNTT',
       ' 823/QĐ_ĐHCNTT', ' 772/QĐ_ĐHCNTT', ' 66 /QĐ-ĐHCNTT',
       ' 222/QĐ-ĐHCNTT', ' 233/QĐ-ĐHCNTT', ' 249/QĐ-ĐHCNTT',
       ' 243/QĐ-ĐHCNTT', ' 384/QĐ-ĐHCNTT', ' 621/QĐ-ĐHCNTT',
       ' 829/QĐ-ĐHCNTT', ' 204 /QĐ-ĐHCNTT', '236', '258',
       ' 298 /QĐ-ĐHCNTT', '306', ' 341 /QĐ-ĐHCNTT', ' 344 /QĐ-ĐHCNTT',
       ' 13/QĐ-ĐHCNTT', ' 443/QĐ-ĐHCNTT', ' 519/QĐ-ĐHCNTT',
       ' 675/QĐ-ĐHCNTT', ' 733/QĐ-ĐHCNTT', ' 761/QĐ-ĐHCNTT',
       ' 811/QĐ-ĐHCNTT', ' 852 /QĐ-ĐHCNTT', ' 960/QĐ-ĐHCNTT',
       ' 214/QĐ-ĐHCNTT', ' 242/QĐ-ĐHCNTT', ' 239/QĐ-ĐHCNTT',
       ' 320/QĐ-ĐHCNTT', ' 310/QĐ-ĐHCNTT', ' 325/QĐ-ĐHCNTT',
       ' 370/QĐ-ĐHCNTT', ' 391/QĐ-ĐHCNTT', ' 430/QĐ-ĐHCNTT',
       ' 542/QĐ-ĐHCNTT', ' 595/QĐ-ĐHCNTT', ' 777/QĐ-ĐHCNTT',
       ' 878/QĐ-ĐHCNTT', ' 955/QĐ-ĐHCNTT', ' 189/QĐ-ĐHCNTT',
       ' 354 /QĐ-ĐHCNTT'], dtype=object)

In [ ]:
df_12_baoluu['ngayqd'].unique()
## Đã OK

array([' 2017-04-14', ' 2017-10-06', ' 2017-11-21', ' 2018-01-19',
       ' 2018-04-10', ' 2018-04-16', ' 2018-04-26', ' 2018-04-23',
       ' 2018-07-03', ' 2018-10-10', ' 2018-12-06', ' 2019-04-12',
       ' 2019-05-02', ' 2019-05-10', ' 2019-05-23', ' 2019-05-29',
       ' 2019-06-05', ' 2019-06-10', ' 2019-01-10', ' 2019-08-01',
       ' 2019-08-26', ' 2019-10-11', ' 2019-10-28', ' 2019-11-05',
       ' 2019-11-22', ' 2019-12-02', ' 2019-12-31', ' 2020-05-18',
       ' 2020-05-27', ' 2020-05-26', ' 2020-06-17', ' 2020-06-15',
       ' 2020-06-19', ' 2020-07-07', ' 2020-07-16', ' 2020-07-31',
       ' 2020-08-31', ' 2020-09-24', ' 2020-11-12', ' 2020-12-07',
       ' 2020-12-30', ' 2021-04-01', ' 2021-06-09'], dtype=object)

In [ ]:
df_12_baoluu.tail(10)

masv  tinhtrang lydo  hocky  namhoc  \
1870  2A1CE9C9XPvAibaEXe86ShnmDByrmwlM6+Yi6Z7F          3  NaN      2    2020   
1871  413BF255XPvAibaEXe+Y0lGmaTa/AIa1iAKZWEji          3  NaN      2    2020   
1872  7695BB67XPvAibaEXe+SI2uhC5CMEVmnYQYHJETH          3  NaN      2    2020   
1873  E3E749B3XPvAibaEXe++R9hzjt1s7yJmiN1bP1O4          3  NaN      2    2020   
1874  7F3EA9C6XPvAibaEXe9kJ+qYUUrp68Mnh6IGKbmi          3  NaN      2    2020   
1875  A2698E63XPvAibaEXe+D5C9Re7EGLdYoD7VxJkyw          3  NaN      2    2020   
1876  B44C3B36XPvAibaEXe/ns/wSuHF6kEcjw3s5q9gL          3  NaN      2    2020   
1877  175475BBXPvAibaEXe8dBqMaN7rGhkux+ZR35bi9          3  NaN      2    2020   
1878  94A79F45XPvAibaEXe+dJypLAXKUXx9tmvinQDNg          3  NaN      2    2020   
1879  B721B939XPvAibaEXe9Q2McsUcINTscwU/8uZNla          3  NaN      2    2020   

                 soqd       ngayqd  
1870   354 /QĐ-ĐHCNTT   2021-06-09  
1871   354 /QĐ-ĐHCNTT   2021-06-09  
1872   354 /QĐ-ĐHCNTT   2021-06-09  
1873   354 /QĐ-ĐHCNTT   2021-06-09  
1874   354 /QĐ-ĐHCNTT   2021-06-09  
1875   354 /QĐ-ĐHCNTT   2021-06-09  
1876   354 /QĐ-ĐHCNTT   2021-06-09  
1877   354 /QĐ-ĐHCNTT   2021-06-09  
1878   354 /QĐ-ĐHCNTT   2021-06-09  
1879   354 /QĐ-ĐHCNTT   2021-06-09

**Nhận xét:** Xóa 2 dòng dữ liệu null và cột index, thu được bảng 12_baoluu gồm 1878 dòng và 7 cột.

## 1'. LÀM SẠCH DỮ LIỆU - V2

In [2]:
import pandas as pd
import numpy as np

In [4]:
df_12_baoluu = pd.read_excel("/content/drive/MyDrive/Khai thác dữ liệu và ứng dụng - CS313.N21/Education_dataset_V2/12.baoluu.xlsx")

In [5]:
df_12_baoluu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1880 entries, 0 to 1879
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          1878 non-null   float64       
 1    masv       1878 non-null   object        
 2    tinhtrang  1878 non-null   float64       
 3    lydo       1878 non-null   object        
 4    hocky      1878 non-null   float64       
 5    namhoc     1878 non-null   float64       
 6    soqd       1878 non-null   object        
 7    ngayqd     1878 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(4), object(3)
memory usage: 117.6+ KB


In [6]:
# Xóa cột id
df_12_baoluu = df_12_baoluu.iloc[:, 1:]

# Chuẩn hóa tên cột (bỏ khoảng trắng trước tên cột)
df_12_baoluu = df_12_baoluu.rename(columns = {' masv': 'masv',
                                              ' tinhtrang': 'tinhtrang',
                                              ' lydo': 'lydo',
                                              ' hocky': 'hocky',
                                              ' namhoc':'namhoc',
                                              ' soqd': 'soqd',
                                              ' ngayqd': 'ngayqd'})
# Kiểm tra
df_12_baoluu.columns

Index(['masv', 'tinhtrang', 'lydo', 'hocky', 'namhoc', 'soqd', 'ngayqd'], dtype='object')

In [7]:
# Xóa 2 dòng không chứa dữ liệu
df_12_baoluu.dropna(axis=0, inplace=True)

In [8]:
df_12_baoluu.shape

(1878, 7)

In [9]:
# # Chuyển kiểu dữ liệu của các cột tinhtrang, hocky, namhoc từ float64 sang int
# df_12_baoluu['tinhtrang'] = df_12_baoluu['tinhtrang'].astype(int)
# df_12_baoluu['hocky'] = df_12_baoluu['hocky'].astype(int)
# df_12_baoluu['namhoc'] = df_12_baoluu['namhoc'].astype(int)

# # Kiểm tra
# df_12_baoluu.info()

In [10]:
df_12_baoluu['lydo'].unique()

array([' TN', ' ',
       ' Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
       ' QĐ điều chỉnh ngành TN', ' Tốt nghiệp'], dtype=object)

In [11]:
# Chuyển ' ' thành nan
# Bỏ các khoảng trắng đầu chuỗi giá trị

df_12_baoluu.replace(' ', np.nan, inplace=True)
df_12_baoluu.replace(' TN', 'Tốt nghiệp', inplace=True)
df_12_baoluu.replace(' Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
                     'Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
                     inplace=True)
df_12_baoluu.replace(' QĐ điều chỉnh ngành TN', 'QĐ điều chỉnh ngành TN', inplace=True)
df_12_baoluu.replace(' Tốt nghiệp', 'Tốt nghiệp', inplace=True)

# Kiểm tra
df_12_baoluu['lydo'].unique()

array(['Tốt nghiệp', nan,
       'Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT',
       'QĐ điều chỉnh ngành TN'], dtype=object)

In [12]:
df_12_baoluu['soqd'].unique()
## Ví soqd với ngayqd có tính chất tương tự nhau nên không cần sử dụng cột soqd
## Suy ra: Không cần chuẩn hóa cột này

array([' 178 /QĐ-ĐHCNTT', ' 713/QD-DHCNTT', ' 713/QĐ-ĐHCNTT',
       ' 823/QĐ_ĐHCNTT', ' 772/QĐ_ĐHCNTT', ' 66 /QĐ-ĐHCNTT',
       ' 222/QĐ-ĐHCNTT', ' 233/QĐ-ĐHCNTT', ' 249/QĐ-ĐHCNTT',
       ' 243/QĐ-ĐHCNTT', ' 384/QĐ-ĐHCNTT', ' 621/QĐ-ĐHCNTT',
       ' 829/QĐ-ĐHCNTT', ' 204 /QĐ-ĐHCNTT', '236', '258',
       ' 298 /QĐ-ĐHCNTT', '306', ' 341 /QĐ-ĐHCNTT', ' 344 /QĐ-ĐHCNTT',
       ' 13/QĐ-ĐHCNTT', ' 443/QĐ-ĐHCNTT', ' 519/QĐ-ĐHCNTT',
       ' 675/QĐ-ĐHCNTT', ' 733/QĐ-ĐHCNTT', ' 761/QĐ-ĐHCNTT',
       ' 811/QĐ-ĐHCNTT', ' 852 /QĐ-ĐHCNTT', ' 960/QĐ-ĐHCNTT',
       ' 214/QĐ-ĐHCNTT', ' 242/QĐ-ĐHCNTT', ' 239/QĐ-ĐHCNTT',
       ' 320/QĐ-ĐHCNTT', ' 310/QĐ-ĐHCNTT', ' 325/QĐ-ĐHCNTT',
       ' 370/QĐ-ĐHCNTT', ' 391/QĐ-ĐHCNTT', ' 430/QĐ-ĐHCNTT',
       ' 542/QĐ-ĐHCNTT', ' 595/QĐ-ĐHCNTT', ' 777/QĐ-ĐHCNTT',
       ' 878/QĐ-ĐHCNTT', ' 955/QĐ-ĐHCNTT', ' 189/QĐ-ĐHCNTT',
       ' 354 /QĐ-ĐHCNTT'], dtype=object)

In [13]:
df_12_baoluu['ngayqd'].unique()
## Đã OK

array(['2017-04-14T00:00:00.000000000', '2017-10-06T00:00:00.000000000',
       '2017-11-21T00:00:00.000000000', '2018-01-19T00:00:00.000000000',
       '2018-04-10T00:00:00.000000000', '2018-04-16T00:00:00.000000000',
       '2018-04-26T00:00:00.000000000', '2018-04-23T00:00:00.000000000',
       '2018-07-03T00:00:00.000000000', '2018-10-10T00:00:00.000000000',
       '2018-12-06T00:00:00.000000000', '2019-04-12T00:00:00.000000000',
       '2019-05-02T00:00:00.000000000', '2019-05-10T00:00:00.000000000',
       '2019-05-23T00:00:00.000000000', '2019-05-29T00:00:00.000000000',
       '2019-06-05T00:00:00.000000000', '2019-06-10T00:00:00.000000000',
       '2019-01-10T00:00:00.000000000', '2019-08-01T00:00:00.000000000',
       '2019-08-26T00:00:00.000000000', '2019-10-11T00:00:00.000000000',
       '2019-10-28T00:00:00.000000000', '2019-11-05T00:00:00.000000000',
       '2019-11-22T00:00:00.000000000', '2019-12-02T00:00:00.000000000',
       '2019-12-31T00:00:00.000000000', '2020-05-18

In [14]:
df_12_baoluu.tail(10)

masv  tinhtrang lydo  hocky  namhoc  \
1870  2A1CE9C9XPvAibaEXe86ShnmDByrmwlM6+Yi6Z7F        3.0  NaN    2.0  2020.0   
1871  413BF255XPvAibaEXe+Y0lGmaTa/AIa1iAKZWEji        3.0  NaN    2.0  2020.0   
1872  7695BB67XPvAibaEXe+SI2uhC5CMEVmnYQYHJETH        3.0  NaN    2.0  2020.0   
1873  E3E749B3XPvAibaEXe++R9hzjt1s7yJmiN1bP1O4        3.0  NaN    2.0  2020.0   
1874  7F3EA9C6XPvAibaEXe9kJ+qYUUrp68Mnh6IGKbmi        3.0  NaN    2.0  2020.0   
1875  A2698E63XPvAibaEXe+D5C9Re7EGLdYoD7VxJkyw        3.0  NaN    2.0  2020.0   
1876  B44C3B36XPvAibaEXe/ns/wSuHF6kEcjw3s5q9gL        3.0  NaN    2.0  2020.0   
1877  175475BBXPvAibaEXe8dBqMaN7rGhkux+ZR35bi9        3.0  NaN    2.0  2020.0   
1878  94A79F45XPvAibaEXe+dJypLAXKUXx9tmvinQDNg        3.0  NaN    2.0  2020.0   
1879  B721B939XPvAibaEXe9Q2McsUcINTscwU/8uZNla        3.0  NaN    2.0  2020.0   

                 soqd     ngayqd  
1870   354 /QĐ-ĐHCNTT 2021-06-09  
1871   354 /QĐ-ĐHCNTT 2021-06-09  
1872   354 /QĐ-ĐHCNTT 2021-06-09  
1873   354 /QĐ-ĐHCNTT 2021-06-09  
1874   354 /QĐ-ĐHCNTT 2021-06-09  
1875   354 /QĐ-ĐHCNTT 2021-06-09  
1876   354 /QĐ-ĐHCNTT 2021-06-09  
1877   354 /QĐ-ĐHCNTT 2021-06-09  
1878   354 /QĐ-ĐHCNTT 2021-06-09  
1879   354 /QĐ-ĐHCNTT 2021-06-09

**Nhận xét:** Xóa 2 dòng dữ liệu null và cột index, thu được bảng 12_baoluu gồm 1878 dòng và 7 cột.

##2. LƯU DỮ LIỆU ĐÃ LÀM SẠCH

In [15]:
# Lưu file csv
df_12_baoluu.to_csv("/content/drive/MyDrive/Khai thác dữ liệu và ứng dụng - CS313.N21/edu_data_v2/12_baoluu.csv",
                    index = False)

##3. PHÂN TÍCH DỮ LIỆU

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
# Đọc dữ liệu từ folder final_edu_data
df = pd.read_csv("/content/drive/MyDrive/Khai thác dữ liệu và ứng dụng - CS313.N21/final_edu_data/12_baoluu.csv")

In [ ]:
df_01_sinhvien = pd.read_csv("/content/drive/MyDrive/Khai thác dữ liệu và ứng dụng - CS313.N21/final_edu_data/01_sinhvien.csv")

In [ ]:
df

masv  tinhtrang        lydo  hocky  \
0     E95E7C6DXPvAibaEXe+1j/AqdkpM22DHf6P99fDJ          3  Tốt nghiệp      2   
1     EBC8A649XPvAibaEXe+fmxcqgvribIboSP1ToKmG          3         NaN      1   
2     2ABBE719XPvAibaEXe9hPQQhQOOPyUJ9WJU9hlCM          3         NaN      1   
3     C307F053XPvAibaEXe+nZBq3b0XEhX8PABYu6iV1          3         NaN      1   
4     F99CE6DEXPvAibaEXe86736b6Ol+/O3ZQU1AW1/C          3         NaN      1   
...                                        ...        ...         ...    ...   
1873  A2698E63XPvAibaEXe+D5C9Re7EGLdYoD7VxJkyw          3         NaN      2   
1874  B44C3B36XPvAibaEXe/ns/wSuHF6kEcjw3s5q9gL          3         NaN      2   
1875  175475BBXPvAibaEXe8dBqMaN7rGhkux+ZR35bi9          3         NaN      2   
1876  94A79F45XPvAibaEXe+dJypLAXKUXx9tmvinQDNg          3         NaN      2   
1877  B721B939XPvAibaEXe9Q2McsUcINTscwU/8uZNla          3         NaN      2   

      namhoc             soqd       ngayqd  
0       2016   178 /QĐ-ĐHCNTT   2017-04-14  
1       2017    713/QD-DHCNTT   2017-10-06  
2       2017    713/QD-DHCNTT   2017-10-06  
3       2017    713/QD-DHCNTT   2017-10-06  
4       2017    713/QD-DHCNTT   2017-10-06  
...      ...              ...          ...  
1873    2020   354 /QĐ-ĐHCNTT   2021-06-09  
1874    2020   354 /QĐ-ĐHCNTT   2021-06-09  
1875    2020   354 /QĐ-ĐHCNTT   2021-06-09  
1876    2020   354 /QĐ-ĐHCNTT   2021-06-09  
1877    2020   354 /QĐ-ĐHCNTT   2021-06-09  

[1878 rows x 7 columns]

#### 3.1. Lý do bảo lưu

In [ ]:
df['lydo'].value_counts()

Tốt nghiệp                                                             199
QĐ điều chỉnh ngành TN                                                   9
Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT      4
Name: lydo, dtype: int64

In [ ]:
df_01 = df[df['lydo']=='QĐ điều chỉnh ngành TN']

In [ ]:
## Những sinh viên này thuộc ngành nào?
result_01 = pd.merge(df_01, df_01_sinhvien, how='inner', on=['masv'])
result_01.loc[:,['masv','lydo','namhoc','masv_tham so dau','namsinh','lopsh']]

masv                    lydo  namhoc  \
0  C47110EFXPvAibaEXe/PGAaNNWvbUF95KJJbQ6Bs  QĐ điều chỉnh ngành TN    2017   
1  DAFCE097XPvAibaEXe+Z7KfNHYLttKsB/kdM5tuZ  QĐ điều chỉnh ngành TN    2017   
2  17ABFF8BXPvAibaEXe9ysTifnGoMwYzIolUfgf2Y  QĐ điều chỉnh ngành TN    2017   
3  4AFE5A8EXPvAibaEXe9MLbI3cCq2BMqcbjt1Z6h0  QĐ điều chỉnh ngành TN    2017   

   masv_tham so dau  namsinh     lopsh  
0              1352   1995.0  ANTT2013  
1              1352   1995.0  ANTT2013  
2              1352   1994.0  ANTT2013  
3              1352   1995.0  ANTT2013

In [ ]:
df_02 = df[df['lydo']=='Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệp chuyển sang hệ CQĐT']
## Những sinh viên này thuộc ngành nào?
result_02 = pd.merge(df_02, df_01_sinhvien, how='inner', on=['masv'])
result_02.loc[:,['masv','lydo','namhoc','masv_tham so dau','namsinh','lopsh']]

masv  \
0  6CDA0ED7XPvAibaEXe+BHZkA9XgGXSS4BHzShGje   
1  EDD0F071XPvAibaEXe+BHZkA9XgGXcEipjLDOZCD   
2  FFE9E452XPvAibaEXe+6MN9FoW2mqXtx1lMTF9+D   

                                                lydo  namhoc  \
0  Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệ...    2017   
1  Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệ...    2017   
2  Chứng chỉ Anh văn không đạt --> CNTN tốt nghiệ...    2017   

   masv_tham so dau  namsinh     lopsh  
0              1352   1995.0  ANTN2013  
1              1352   1995.0  ANTN2013  
2              1352   1995.0  ANTN2013

In [ ]:
df_03 = df[df['lydo']=='Tốt nghiệp']
## Tại sao lí do bảo lưu lại là tốt nghiệp?
result_03 = pd.merge(df_03, df_01_sinhvien, how='inner', on=['masv'])
result_03.loc[:,['masv','lydo','namhoc','masv_tham so dau','namsinh','lopsh']]

masv        lydo  namhoc  \
0  E95E7C6DXPvAibaEXe+1j/AqdkpM22DHf6P99fDJ  Tốt nghiệp    2016   
1  FB623556XPvAibaEXe8Y+/3uR2PkTNbh/Fe4KQiS  Tốt nghiệp    2020   
2  242ADB97XPvAibaEXe9ysTifnGoMwfpn00Jzt4a2  Tốt nghiệp    2020   
3  08BCA73BXPvAibaEXe8VekUw4ELehgGhnY+JjnAc  Tốt nghiệp    2020   
4  0B468FE1XPvAibaEXe+fmxcqgvribL7shgQ1FhJv  Tốt nghiệp    2020   
5  12366023XPvAibaEXe+PhKfQb0+XJ9/8933Iy6CC  Tốt nghiệp    2020   
6  CB36ADA7XPvAibaEXe/caEx4rNa5qRJOcczZFsXE  Tốt nghiệp    2020   
7  F0F2F362XPvAibaEXe9upvU5DvuC9vpGmW+2YsWe  Tốt nghiệp    2020   

   masv_tham so dau  namsinh     lopsh  
0              1352   1994.0  MMTT2013  
1              1352   1995.0  KHMT0001  
2              1352   1995.0  KHMT0001  
3              1352   1995.0  KTPM0001  
4              1352   1995.0  KTPM0001  
5              1352   1995.0  MMTT0001  
6              1352   1995.0  MMTT0001  
7              1352   1995.0  KTMT0001

In [ ]:
print("---------Namhoc-------")
print(result_03['namhoc'].value_counts(), '\n')

print("---------Namhoc-------")
print(result_03['masv_tham so dau'].value_counts(), '\n')
## Chỉ có 8 sinh viên có mssv tham chiếu đến bảng 01_sinhvien

---------Namhoc-------
2020    198
2016      1
Name: namhoc, dtype: int64 

---------Namhoc-------
1352.0    8
Name: masv_tham so dau, dtype: int64 



####3.2. Số lượng sv bảo lưu từng năm

In [ ]:
df['namhoc'].value_counts()

2020    726
2018    441
2019    358
2017    352
2016      1
Name: namhoc, dtype: int64

In [ ]:
df['hocky'].value_counts()

1    909
2    891
3     78
Name: hocky, dtype: int64

In [ ]:
df[df['hocky']==3]
## Tại sao có học kì 3?

masv  tinhtrang lydo  hocky  namhoc  \
716  981C7989XPvAibaEXe9P07hcrvmhCcxHupzWlaJ2          3  NaN      3    2018   
717  FDE6070DXPvAibaEXe8zt9d6Yn9pDDazARoQk6ZP          3  NaN      3    2018   
718  4785283EXPvAibaEXe8ptkRJyotmH0AxIrbY3Kvy          3  NaN      3    2018   
719  AAB36911XPvAibaEXe9ETemEVgJcURSE4QiYowQq          3  NaN      3    2018   
720  3B1B73AEXPvAibaEXe/y2t7DVG8TDass+/GF6+ZL          3  NaN      3    2018   
..                                        ...        ...  ...    ...     ...   
789  2D99E412XPvAibaEXe8VziDgYa0u7pmPv7ThgrVQ          3  NaN      3    2018   
790  1E3B0973XPvAibaEXe8dEe4TA8wgHxwHmaQUgsiZ          3  NaN      3    2018   
791  19FE69E6XPvAibaEXe8IlpBhypyXxnoaAYkrfDkq          3  NaN      3    2018   
792  28807FE1XPvAibaEXe8k9aD0YXzUPpNUUND3x7rK          3  NaN      3    2018   
793  04F87408XPvAibaEXe80b1fr0fH/JuobsigdDkbt          3  NaN      3    2018   

               soqd       ngayqd  
716   443/QĐ-ĐHCNTT   2019-08-01  
717   443/QĐ-ĐHCNTT   2019-08-01  
718   443/QĐ-ĐHCNTT   2019-08-01  
719   443/QĐ-ĐHCNTT   2019-08-01  
720   443/QĐ-ĐHCNTT   2019-08-01  
..              ...          ...  
789   443/QĐ-ĐHCNTT   2019-08-01  
790   443/QĐ-ĐHCNTT   2019-08-01  
791   443/QĐ-ĐHCNTT   2019-08-01  
792   443/QĐ-ĐHCNTT   2019-08-01  
793   519/QĐ-ĐHCNTT   2019-08-26  

[78 rows x 7 columns]

In [ ]:
df[df['hocky']==3]['ngayqd'].value_counts()

 2019-08-01    77
 2019-08-26     1
Name: ngayqd, dtype: int64

####3.3. Sinh viên thường bảo lưu vào năm thứ mấy của CTĐT

In [ ]:
## Cần kết với bảng sinhvien để biết sinh viên thuộc khóa mấy
result_04 = pd.merge(df, df_01_sinhvien, how='left', on=['masv'])
result_04.loc[:, ['']]